Unhelpful conflict messages

### *** Please add your unhelpful conflict message to this [discussion](https://github.com/ocaml/opam/discussions/5130) ***
---

While trying to install `expect.0.0.6` on OCaml 4.12:
```
[ERROR] Package conflict!
No solution found, exiting
  * No agreement on the version of ocaml:
    - (invariant) -> ocaml-variants = 4.12.0+trunk -> ocaml >= 4.12.0
    - expect >= 0.0.6 -> batteries -> ocaml < 4.00.0
    You can temporarily relax the switch invariant with `--update-invariant'
```
I would have expected:
```
    - expect >= 0.0.6 -> batteries -> ocaml < 4.12.0
```
as the latest version of `batteries` indicates.

Similar here for `ppx_expect.v0.14.0`:
```
[ERROR] Package conflict!
No solution found, exiting
  * No agreement on the version of ocaml:
    - (invariant) -> ocaml-variants = 4.12.0+trunk -> ocaml >= 4.12.0
    - ppx_expect >= v0.14.0 -> ppxlib >= 0.11.0 -> ocaml < 4.10.0
    You can temporarily relax the switch invariant with `--update-invariant'
```
I would have expected an extra step:
```
    - ppx_expect >= v0.14.0 -> ppxlib >= 0.11.0 -> ocaml-migrate-parsetree >= 2.0.0 -> ocaml < 4.12
```

I have another really surprising and different one here:
```
$ opam switch show
4.12
$ opam install gen_js_api.1.0.6
[ERROR] Package conflict!
Sorry, no solution found: there seems to be a problem with your request.

No solution found, exiting
```

More testing with dose 5.0.1:
```
opam@22cdd2d0a793:~$ opam-2.0 install pgocaml_ppx.4.2.2
The following dependencies couldn't be met:
  - pgocaml_ppx -> ocaml-migrate-parsetree < 2.0.0 -> ocaml < 4.13
      base of this switch (use `--unlock-base' to force)
  - pgocaml_ppx -> ocaml-migrate-parsetree < 2.0.0 -> dune < 2.0 -> ocaml < 4.12
      base of this switch (use `--unlock-base' to force)

No solution found, exiting
opam@22cdd2d0a793:~$ opam-2.1 install pgocaml_ppx.4.2.2
[ERROR] Package conflict!
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 -> ocaml-migrate-parsetree < 2.0.0 -> ocaml-variants = 4.08.0+beta2
    not available because the package is pinned to version 4.13.0+trunk
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 -> ocaml-migrate-parsetree < 2.0.0 -> ocaml-variants = 4.08.0+beta3
    not available because the package is pinned to version 4.13.0+trunk

No solution found, exiting
```
Switching to master with dose 7.0.0, there seems to be no differences with opam 2.1:
```
$ opam --version
2.2.0~alpha~dev
$ opam install pgocaml_ppx.4.2.2
[ERROR] Package conflict!
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 -> ocaml-migrate-parsetree < 2.0.0 -> ocaml-variants = 4.08.0+beta2
    not available because the package is pinned to version 4.13.0+trunk
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 -> ocaml-migrate-parsetree < 2.0.0 -> ocaml-variants = 4.08.0+beta3
    not available because the package is pinned to version 4.13.0+trunk

No solution found, exiting
```
I also tested on a macOS/arm64 with dose 7.0.0 and it seems that the constraint solver does not like having "available constraints" in the compiler packages:
```
$ opam install pgocaml_ppx.4.2.2
[ERROR] Package conflict!
  * No agreement on the version of ocaml:
    - (invariant) → ocaml-variants = 4.13.0+trunk → ocaml = 4.13.0
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.11
    You can temporarily relax the switch invariant with `--update-invariant'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler < 3.07+1 | ocaml-system
      < 3.07+1 | ocaml-variants < 3.8~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions: 'sys-ocaml-version = "3.07"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.07+1 | ocaml-system
      = 3.07+1 | ocaml-variants < 3.8~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions: 'sys-ocaml-version = "3.07+1"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.07+2 | ocaml-system
      = 3.07+2 | ocaml-variants < 3.8~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions: 'sys-ocaml-version = "3.07+2"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.08.0 | ocaml-system
      >= 3.08.0 | ocaml-variants < 3.08.1~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.08.0"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.08.1 | ocaml-system
      >= 3.08.1 | ocaml-variants < 3.08.2~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.08.1"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.08.2 | ocaml-system
      >= 3.08.2 | ocaml-variants < 3.08.3~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.08.2"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.08.3 | ocaml-system
      >= 3.08.3 | ocaml-variants < 3.08.4~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.08.3"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.08.4 | ocaml-system
      >= 3.08.4 | ocaml-variants < 3.08.5~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.08.4"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.09.0 | ocaml-system
      >= 3.09.0 | ocaml-variants < 3.09.1~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.09.0"'
    no matching version
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.09.1 | ocaml-system
      >= 3.09.1 | ocaml-variants < 3.09.2~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.09.1"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.09.2 | ocaml-system
      >= 3.09.2 | ocaml-variants < 3.09.3~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.09.2"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.09.3 | ocaml-system
      >= 3.09.3 | ocaml-variants < 3.09.4~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.09.3"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.10.0 | ocaml-system
      >= 3.10.0 | ocaml-variants < 3.10.1~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.10.0"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.10.1 | ocaml-system
      >= 3.10.1 | ocaml-variants < 3.10.2~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.10.1"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.10.2 | ocaml-system
      >= 3.10.2 | ocaml-variants < 3.10.3~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.10.2"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.11.0 | ocaml-system
      >= 3.11.0 | ocaml-variants < 3.11.1~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.11.0"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.11.1 | ocaml-system
      >= 3.11.1 | ocaml-variants < 3.11.2~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.11.1"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.11.2 | ocaml-system
      >= 3.11.2 | ocaml-variants < 3.11.3~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.11.2"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.12.0 | ocaml-system
      >= 3.12.0 | ocaml-variants < 3.12.1~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.12.0"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 3.12.1 | ocaml-system
      >= 3.12.1 | ocaml-variants < 3.12.2~
    unmet availability conditions: 'arch != "arm64" & arch != "arm32" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "3.12.1"'
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 4.00.0 | ocaml-system
      >= 4.00.0 | ocaml-variants < 4.00.1~
    unmet availability conditions: 'arch != "arm64" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.00.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 4.00.1 | ocaml-system
      >= 4.00.1 | ocaml-variants < 4.00.2~
    unmet availability conditions: 'arch != "arm64" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.00.1"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 4.01.0 | ocaml-system
      >= 4.01.0 | ocaml-variants < 4.01.1~
    unmet availability conditions: 'arch != "arm64" & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.01.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml = 4.02.0 → ocaml-base-compiler = 4.02.0 | ocaml-system
      >= 4.02.0 | ocaml-variants < 4.02.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64") & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.02.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 4.02.1 | ocaml-system
      >= 4.02.1 | ocaml-variants < 4.02.2~
    unmet availability conditions: '!(os = "macos" & arch = "arm64") & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.02.1"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 4.02.2 | ocaml-system
      >= 4.02.2 | ocaml-variants < 4.02.3~
    unmet availability conditions: '!(os = "macos" & arch = "arm64") & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.02.2"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 4.02.3 | ocaml-system
      >= 4.02.3 | ocaml-variants < 4.02.4~
    unmet availability conditions: '!(os = "macos" & arch = "arm64") & arch != "ppc64"'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.02.3"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.03.0 → ocaml-base-compiler = 4.02.4 | ocaml-system
      >= 4.02.4 | ocaml-variants < 4.02.5~
    no matching version
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.04 → ocaml-base-compiler = 4.03.0 | ocaml-system
      >= 4.03.0 | ocaml-variants < 4.03.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.03.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.04 → ocaml-base-compiler = 4.03.1 | ocaml-system
      >= 4.03.1 | ocaml-variants < 4.03.2~
    no matching version
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.05 → ocaml-base-compiler = 4.04.0 | ocaml-system
      >= 4.04.0 | ocaml-variants < 4.04.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.04.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.05 → ocaml-base-compiler = 4.04.1 | ocaml-system
      >= 4.04.1 | ocaml-variants < 4.04.2~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.04.1"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.05 → ocaml-base-compiler = 4.04.2 | ocaml-system
      >= 4.04.2 | ocaml-variants < 4.04.3~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.04.2"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.05 → ocaml-base-compiler = 4.04.3 | ocaml-system
      >= 4.04.3 | ocaml-variants < 4.04.4~
    no matching version
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.06.0 → ocaml-base-compiler = 4.05.0 | ocaml-system
      >= 4.05.0 | ocaml-variants < 4.05.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.05.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.06.0 → ocaml-base-compiler = 4.05.1 | ocaml-system
      >= 4.05.1 | ocaml-variants < 4.05.2~
    no matching version
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.08 → ocaml-base-compiler = 4.06.0 | ocaml-system
      >= 4.06.0 | ocaml-variants < 4.06.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.06.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.08 → ocaml-base-compiler = 4.06.1 | ocaml-system
      >= 4.06.1 | ocaml-variants < 4.06.2~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.06.1"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.08 → ocaml-base-compiler = 4.06.2 | ocaml-system
      >= 4.06.2 | ocaml-variants < 4.06.3~
    no matching version
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.08 → ocaml-base-compiler = 4.07.0 | ocaml-system
      >= 4.07.0 | ocaml-variants < 4.07.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.07.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.08 → ocaml-base-compiler = 4.07.1 | ocaml-system
      >= 4.07.1 | ocaml-variants < 4.07.2~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.07.1"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.08 → ocaml-base-compiler = 4.07.2 | ocaml-system
      >= 4.07.2 | ocaml-variants < 4.07.3~
    no matching version
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.10 → ocaml-base-compiler = 4.08.0 | ocaml-system
      >= 4.08.0 | ocaml-variants < 4.08.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.08.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.10 → ocaml-base-compiler = 4.08.1 | ocaml-system
      >= 4.08.1 | ocaml-variants < 4.08.2~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.08.1"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.10 → ocaml-base-compiler = 4.08.2 | ocaml-system
      >= 4.08.2 | ocaml-variants < 4.08.3~
    no matching version
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.10 → ocaml-base-compiler = 4.09.0 | ocaml-system
      >= 4.09.0 | ocaml-variants < 4.09.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.09.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.10 → ocaml-base-compiler = 4.09.1 | ocaml-system
      >= 4.09.1 | ocaml-variants < 4.09.2~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.09.1"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.10 → ocaml-base-compiler = 4.09.2 | ocaml-system
      >= 4.09.2 | ocaml-variants < 4.09.3~
    no matching version
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.12.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.11 → ocaml-base-compiler = 4.10.0 | ocaml-system
      >= 4.10.0 | ocaml-variants < 4.10.1~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.10.0"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'
  * Missing dependency:
    - pgocaml_ppx = 4.2.2 → ppx_tools → ocaml < 4.11 → ocaml-base-compiler = 4.10.1 | ocaml-system
      >= 4.10.1 | ocaml-variants < 4.10.2~
    unmet availability conditions: '!(os = "macos" & arch = "arm64")'
    unmet availability conditions, e.g. 'sys-ocaml-version = "4.10.1"'
    unmet availability conditions, e.g. '!(os = "macos" & arch = "arm64")'

No solution found, exiting
```
So overall the issue won't be resolved by upgrading dose. It is an endemic issue since opam 2.1 (as seen above by the clear difference between the conflict message in 2.0 and 2.1)

Hi, currently I try to use `opam.2.1.0` with a certain version of `unipi` available here: https://github.com/roburio/unipi/pull/4 without `pin` and `opam` shows this error:
```sh
$ git clone https://github.com/dinosaure/unipi
$ cd unipi
$ git checkout with-paf-opam
$ mirage configure
$ make depends
opam pin add -k path --no-action --yes mirage-unikernel-unipi-unix . && opam depext --yes --update mirage-unikernel-unipi-unix ; opam pin remove --no-action mirage-unikernel-unipi-unix
Package mirage-unikernel-unipi-unix does not exist, create as a NEW package? [Y/n] y
[mirage-unikernel-unipi-unix.~dev] synchronised (file:///home/dinosaure/dev/unipi)
mirage-unikernel-unipi-unix is now pinned to file:///home/dinosaure/dev/unipi (version ~dev)
opam: unknown command `depext'.
Usage: opam COMMAND ...
Try `opam --help' for more information.
Ok, mirage-unikernel-unipi-unix is no longer pinned to file:///home/dinosaure/dev/unipi (version ~dev)
opam install -y --deps-only .
[ERROR] Package conflict!
Sorry, no solution found: there seems to be a problem with your request.

No solution found, exiting
make: *** [Makefile:15 : depends] Erreur 20
```

So I believe that something is wrong about dependencies but I don't have any clues 😕 . Do you know an option to explain where is the error?

Same issue with opam-repository at `a5d7cdc0c91452b0aef4fa71c331ee5237f6dddd` and https://github.com/bikallem/resto/commit/e725f56dd2d5ca0148272d754a0f29ffe76f94e8 only gets us:
```
[ERROR] Package conflict!
Sorry, no solution found: there seems to be a problem with your request.
```

--- a/master_changes.md
+++ b/master_changes.md
@@ -160,6 +160,7 @@ users)
 ## Client
   * Check whether the repository might need updating more often [#4935 @kit-ty-kate]
   * ✘ It is no longer possible to process actions on packages that depend on a package that was removed upstream [#4969 @altgr]
+  * Fix (at least some of the) empty conflict explanations [#4982 @kit-ty-kate]
 
 ## Internal
   * Add license and lowerbounds to opam files [#4714 @kit-ty-kate]

--- a/src/solver/opamCudf.ml
+++ b/src/solver/opamCudf.ml
@@ -854,7 +854,10 @@ let extract_explanations packages cudfnv2opam reasons : explanation list =
               Map.update pkg (CS.union chain) CS.empty acc
             in
             let ds = get deps p in
-            let dsc = ds %% all_conflicting in
+            let dsc = match Hashtbl.length missing with
+              | 0 -> ds (* Hack to fix https://github.com/ocaml/opam/issues/4373. We should try to do better at some point *)
+              | _ -> ds %% all_conflicting
+            in
             if not (Set.is_empty dsc) then
               dsc ++ seen1, Set.fold append_to_chains (dsc -- seen1) new_chains
             else

--- a/tests/reftests/conflict-resto.test
+++ b/tests/reftests/conflict-resto.test
@@ -388,7 +388,7 @@ Faking installation of ocaml-config.2
 Faking installation of ocaml.4.12.0
 Faking installation of ocaml-options-vanilla.1
 Done.
-### opam pin resto/ --with-version=0.6.1 --show --yes | unordered
+### OPAMVAR_os=linux opam pin resto/ --with-version=0.6.1 --show --yes | unordered
 This will pin the following packages: ezresto-directory, ezresto, resto-acl, resto-client, resto-cohttp-client, resto-cohttp-self-serving-client, resto-cohttp-server, resto-cohttp, resto-directory, resto-json, resto-lwt, resto-self-serving-client, resto-server, resto. Continue? [Y/n] y
 ezresto-directory is now pinned to file://${BASEDIR}/resto (version 0.6.1)
 ezresto is now pinned to file://${BASEDIR}/resto (version 0.6.1)

--- a/tests/reftests/conflict-resto.test
+++ b/tests/reftests/conflict-resto.test
@@ -411,8 +411,8 @@ resto is now pinned to file://${BASEDIR}/resto (version 0.6.1)
 
 [ERROR] Package conflict!
   * No agreement on the version of conduit-lwt-unix:
-    - resto-cohttp-server -> cohttp-lwt-unix >= 4.0.0 -> conduit-lwt-unix < 5.0.0
-    - resto-cohttp-server -> conduit-lwt-unix >= 5.0.0
+    - resto-lwt -> cohttp-lwt-unix >= 4.0.0 -> conduit-lwt-unix < 5.0.0
+    - resto-lwt -> conduit-lwt-unix >= 5.0.0
 
 [NOTE] Pinning command successful, but your installed packages may be out of sync.
 # Return code 20 #

--- a/tests/reftests/conflict-resto.test
+++ b/tests/reftests/conflict-resto.test
@@ -410,7 +410,9 @@ resto-server is now pinned to file://${BASEDIR}/resto (version 0.6.1)
 resto is now pinned to file://${BASEDIR}/resto (version 0.6.1)
 
 [ERROR] Package conflict!
-Sorry, no solution found: there seems to be a problem with your request.
+  * No agreement on the version of conduit-lwt-unix:
+    - resto-cohttp-server -> cohttp-lwt-unix >= 4.0.0 -> conduit-lwt-unix < 5.0.0
+    - resto-cohttp-server -> conduit-lwt-unix >= 5.0.0
 
 [NOTE] Pinning command successful, but your installed packages may be out of sync.
 # Return code 20 #
@@ -434,7 +436,9 @@ conflicts: "a" {< "2"}
 
 ### opam install b c --show
 [ERROR] Package conflict!
-Sorry, no solution found: there seems to be a problem with your request.
+  * Incompatible packages:
+    - b -> a < 2
+    - c
 
 No solution found, exiting
 # Return code 20 #

